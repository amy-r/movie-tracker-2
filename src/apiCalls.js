import { key } from './key';

export const test = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1`;

export const fetchParse = async url => {
  try {
    const response = await fetch(url);
    const parsed = await response.json();
    return parsed;
  } catch (error) {
    return 'fetchParse Err';
  }
};

export const movieCleaner = movies => {
  if (movies === null) {
    return undefined;
  }
  return movies.results.map(movie => {
    const {
      title,
      poster_path,
      id,
      release_date,
      vote_average,
      overview
    } = movie;
    return {
      title,
      poster_path,
      movie_id: id,
      release_date,
      vote_average,
      overview
    };
  });
};

export const logIn = async user => {
  try {
    const response = await fetch('/api/users', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed.data;
    }
  } catch (error) {
    return "Error in logIn";
  }
};

export const createUser = async user => {
  try {
    const response = await fetch('/api/users/new', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return "Error in createUser";
  }
};

export const addFavorite = async (movie) => {
  try {
     const response = await fetch('api/users/favorites/new', { 
      method: 'POST', 
      body: JSON.stringify(movie),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in addFavorite';
  }   
}

export const removeFavorite = async (movie) => {
  const { movie_id, user_id } = movie;
  try {
     const response = await fetch(`api/users/${user_id}/favorites/${movie_id}`, { 
      method: 'DELETE', 
      body: JSON.stringify({ movie_id, user_id }),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if(response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in removeFavorite';
  }   
}

export const fetchAllFavorites = async (userId) => {
  try {
    const response = await fetch(`api/users/${userId}/favorites`)
    const favorites = await response.json()
    return favorites.data
  } catch(error) {
      return 'Error in fetchAllFavorites';
  }
  
}