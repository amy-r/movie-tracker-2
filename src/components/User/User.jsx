import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { logIn } from '../../actions/actions';
import * as api from '../../apiCalls';
import PropTypes from 'prop-types';
import './User.css';

export class User extends Component {
  constructor() {
    super();
    this.state = {
      password: '',
      email: ''
    };
  }

  handleInput = event => {
    const { value, name } = event.target;
    this.setState({ [name]: value });
  };

  handleSubmit = async event => {
    event.preventDefault();
    const user = await api.logIn(this.state);
    if (user) {
      this.props.logIn(user);
      this.props.history.push('/');
    } else {
      alert('WRONG PASSWORD');
      this.setState({ email: '', password: '' });
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="email">Email: </label>
          <input
            type="text"
            id="email"
            name="email"
            onChange={this.handleInput}
            value={this.state.email}
          />
          <label htmlFor="password">Password: </label>
          <input
            type="text"
            id="password"
            name="password"
            value={this.state.password}
            onChange={this.handleInput}
          />
          <input type="submit" />
        </form>
      </div>
    );
  }
}

const mapStateToProps = store => ({});

const mapDispatchToProps = dispatch => ({
  logIn: user => dispatch(logIn(user))
});

User.propTypes = {
  logIn: PropTypes.func,
  history: PropTypes.object
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(User));
