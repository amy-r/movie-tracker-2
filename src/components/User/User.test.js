/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { User, mapDispatchToProps } from './User';
import * as api from '../../apiCalls';

describe('USER', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<User />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state ', () => {
    expect(wrapper.state().password).toEqual('');
    expect(wrapper.state().email).toEqual('');
  });

  describe('handleInput', () => {
    it('should setState of email', () => {
      const expected = 'howdy';
      const mockEvent = {
        preventDefault: jest.fn(),
        target: {
          name: 'email',
          value: 'howdy'
        }
      };
      wrapper.instance().handleInput(mockEvent);
      wrapper.update();
      expect(wrapper.state().email).toEqual(expected);
    });

    it('should setState of password', () => {
      const expected = 'whats up doc';
      const mockEvent = {
        preventDefault: jest.fn(),
        target: { name: 'password', value: 'whats up doc' }
      };
      wrapper.instance().handleInput(mockEvent);
      wrapper.update();
      expect(wrapper.state().password).toEqual(expected);
    });
  });

  describe('handleSubmit', () => {
    it('should prevent default', async () => {
      wrapper = shallow(<User logIn={jest.fn()} history={[]} />);
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().handleSubmit(mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled();
    });

    it('should login the user by sending state to api.logIn', () => {
      wrapper = shallow(<User logIn={jest.fn()} history={[]} />);
      api.logIn = jest.fn();
      const mockState = { password: 'suh', email: 'dude@dude.com' };
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.setState(mockState);
      wrapper.instance().handleSubmit(mockEvent);
      expect(api.logIn).toHaveBeenCalledWith(mockState);
    });

    it.skip('should take the response from DB and send to store', () => {
      wrapper = shallow(<User logIn={jest.fn()} history={[]} />);
      api.logIn = jest
        .fn()
        .mockImplementation(() =>
          Promise.resolve({
            ok: true,
            json: () =>
              Promise.resolve({
                data: {
                  id: 0,
                  password: 'suh',
                  email: 'dude@dude.com',
                  name: 'jhun'
                }
              })
          })
        );
      const mockState = { password: 'suh', email: 'dude@dude.com' };
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.setState(mockState);
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.props().logIn).toHaveBeenCalled();
    });
  });

  describe('userReject', () => {
    it('should reset state to default if wrong password', () => {
      const expected = { email: '', password: '' };
      wrapper.setState({ email: 'bar', password: 'foo' });
      wrapper.instance().userReject();
      expect(wrapper.state()).toEqual(expected);
    });
  });

  describe('MDTP', () => {
    it('should call the dispatch function on MDTP', () => {
      const mockDispatch = jest.fn();
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.logIn();
      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});
