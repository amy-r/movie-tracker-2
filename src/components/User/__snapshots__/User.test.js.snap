// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USER should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <User />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <label
                htmlFor="email"
        >
                Email: 
        </label>
        <input
                id="email"
                name="email"
                onChange={[Function]}
                type="text"
                value=""
        />
        <label
                htmlFor="password"
        >
                Password: 
        </label>
        <input
                id="password"
                name="password"
                onChange={[Function]}
                type="text"
                value=""
        />
        <input
                type="submit"
        />
</form>,
      "className": "User",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="email"
>
            Email: 
</label>,
          <input
            id="email"
            name="email"
            onChange={[Function]}
            type="text"
            value=""
/>,
          <label
            htmlFor="password"
>
            Password: 
</label>,
          <input
            id="password"
            name="password"
            onChange={[Function]}
            type="text"
            value=""
/>,
          <input
            type="submit"
/>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Email: ",
            "htmlFor": "email",
          },
          "ref": null,
          "rendered": "Email: ",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "email",
            "name": "email",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Password: ",
            "htmlFor": "password",
          },
          "ref": null,
          "rendered": "Password: ",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "password",
            "name": "password",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <label
                    htmlFor="email"
          >
                    Email: 
          </label>
          <input
                    id="email"
                    name="email"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <label
                    htmlFor="password"
          >
                    Password: 
          </label>
          <input
                    id="password"
                    name="password"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <input
                    type="submit"
          />
</form>,
        "className": "User",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              htmlFor="email"
>
              Email: 
</label>,
            <input
              id="email"
              name="email"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <label
              htmlFor="password"
>
              Password: 
</label>,
            <input
              id="password"
              name="password"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <input
              type="submit"
/>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Email: ",
              "htmlFor": "email",
            },
            "ref": null,
            "rendered": "Email: ",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "email",
              "name": "email",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Password: ",
              "htmlFor": "password",
            },
            "ref": null,
            "rendered": "Password: ",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "password",
              "name": "password",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
